File: python/packages/jumpstarter-cli-common/jumpstarter_cli_common/oidc.py
import json
import os
import ssl
import time
import warnings
from dataclasses import dataclass
from functools import wraps
from typing import ClassVar

import aiohttp
import anyio
import certifi
import click
from aiohttp import web
from anyio import create_memory_object_stream
from anyio.to_thread import run_sync

# Authlib still pulls deprecated authlib.jose on first OAuth2Session use, not only at
# import time; a transient catch_warnings() around the import does not suppress that.
warnings.filterwarnings(
    "ignore",
    message=r"authlib\.jose module is deprecated.*",
    module=r"authlib\..*",
)

# When the user opts into --insecure-tls, we set verify=False on the requests session.
# urllib3 emits InsecureRequestWarning for every such request; suppress it since the
# user has already acknowledged the risk.
warnings.filterwarnings(
    "ignore",
    message=r"Unverified HTTPS request is being made.*",
    module=r"urllib3\..*",
)
from authlib.integrations.requests_client import OAuth2Session
from joserfc.jws import extract_compact
from yarl import URL

from jumpstarter.config.env import JMP_OIDC_CALLBACK_PORT, JMP_OIDC_DEVICE_FLOW

DEVICE_FLOW_POLL_INTERVAL = 5
DEVICE_FLOW_GRANT_TYPE = "urn:ietf:params:oauth:grant-type:device_code"


def _get_ssl_context() -> ssl.SSLContext:
    """Create an SSL context that respects SSL_CERT_FILE environment variable."""
    ssl_ctx = ssl.create_default_context()
    ca_file = os.environ.get("SSL_CERT_FILE") or certifi.where()
    ssl_ctx.load_verify_locations(ca_file)
    return ssl_ctx


def opt_oidc(f):
    @click.option("--issuer", help="OIDC issuer")
    @click.option("--client-id", "client_id", help="OIDC client id", default="jumpstarter-cli")
    @click.option("--token", help="OIDC access token")
    @click.option("--username", help="OIDC username")
    @click.option("--password", help="OIDC password")
    @click.option("--connector-id", "connector_id", help="OIDC token exchange connector id (Dex specific)")
    @click.option(
        "--callback-port",
        "callback_port",
        type=click.IntRange(0, 65535),
        default=None,
        help="Port for OIDC callback server (0=random port)",
    )
    @click.option(
        "--offline-access/--no-offline-access",
        default=True,
        help="Request offline_access scope (refresh token)",
    )
    @click.option(
        "--device-flow/--no-device-flow",
        "device_flow",
        default=None,
        help="Use OAuth 2.0 Device Authorization Grant (auto-detected in Dev Spaces)",
    )
    @wraps(f)
    def wrapper(*args, **kwds):
        return f(*args, **kwds)

    return wrapper


@dataclass(kw_only=True)
class Config:
    issuer: str
    client_id: str
    offline_access: bool = False
    insecure_tls: bool = False
    scope: ClassVar[list[str]] = ["openid", "profile"]

    async def configuration(self):
        ssl_context: ssl.SSLContext | bool = False if self.insecure_tls else _get_ssl_context()
        connector = aiohttp.TCPConnector(ssl=ssl_context)
        async with aiohttp.ClientSession(connector=connector) as session:
            async with session.get(
                URL(self.issuer).joinpath(".well-known", "openid-configuration"),
                raise_for_status=True,
            ) as response:
                return await response.json()

    def _scopes(self) -> list[str]:
        if self.offline_access:
            return [*self.scope, "offline_access"]
        return list(self.scope)

    def client(self, **kwargs):
        session = OAuth2Session(client_id=self.client_id, scope=self._scopes(), **kwargs)
        session.verify = False if self.insecure_tls else (os.environ.get("SSL_CERT_FILE") or certifi.where())
        return session

    async def token_exchange_grant(self, token: str, **kwargs):
        config = await self.configuration()

        client = self.client()

        return await run_sync(
            lambda: client.fetch_token(
                config["token_endpoint"],
                grant_type="urn:ietf:params:oauth:grant-type:token-exchange",
                requested_token_type="urn:ietf:params:oauth:token-type:access_token",
                subject_token_type="urn:ietf:params:oauth:token-type:id_token",
                subject_token=token,
                **kwargs,
            )
        )

    async def refresh_token_grant(self, refresh_token: str):
        config = await self.configuration()

        client = self.client()

        return await run_sync(
            lambda: client.fetch_token(
                config["token_endpoint"],
                grant_type="refresh_token",
                refresh_token=refresh_token,
            )
        )

    async def password_grant(self, username: str, password: str):
        config = await self.configuration()

        client = self.client()

        return await run_sync(
            lambda: client.fetch_token(
                config["token_endpoint"], grant_type="password", username=username, password=password
            )
        )

    async def authorization_code_grant(self, callback_port: int | None = None, prompt: str | None = None):
        config = await self.configuration()

        # Use provided port, fall back to env var, then default to 0 (OS picks)
        if callback_port is not None:
            port = callback_port
        else:
            env_value = os.environ.get(JMP_OIDC_CALLBACK_PORT)
            if env_value is None:
                port = 0
            elif env_value.isdigit() and int(env_value) <= 65535:
                port = int(env_value)
            else:
                raise click.ClickException(f'Invalid {JMP_OIDC_CALLBACK_PORT} "{env_value}": must be a valid port')

        tx, rx = create_memory_object_stream()

        async def callback(request):
            await tx.send(str(request.url))
            return web.Response(text="Login successful, you can close this page")

        app = web.Application()
        app.add_routes([web.get("/callback", callback)])

        runner = web.AppRunner(app, access_log=None)
        await runner.setup()

        site = web.TCPSite(runner, "localhost", port)
        try:
            await site.start()
        except OSError as e:
            await runner.cleanup()
            raise click.ClickException(f"Failed to start callback server on port {port}: {e}") from None

        redirect_uri = "http://localhost:%d/callback" % site._server.sockets[0].getsockname()[1]

        client = self.client(redirect_uri=redirect_uri)

        # Add prompt parameter if force requested
        auth_params = {}
        if prompt:
            auth_params["prompt"] = prompt

        uri, state = client.create_authorization_url(config["authorization_endpoint"], **auth_params)

        print("Please open the URL in browser: ", uri)

        authorization_response = await rx.receive()

        await site.stop()
        await runner.cleanup()

        return await run_sync(
            lambda: client.fetch_token(config["token_endpoint"], authorization_response=authorization_response)
        )

    async def device_code_grant(self):
        config = await self.configuration()

        device_endpoint = config.get("device_authorization_endpoint")
        if not device_endpoint:
            raise click.ClickException(
                "OIDC provider does not support Device Authorization Grant. "
                "Ensure 'oauth2DeviceAuthorizationGrantEnabled' is enabled on the OIDC client."
            )

        ssl_context: ssl.SSLContext | bool = False if self.insecure_tls else _get_ssl_context()
        connector = aiohttp.TCPConnector(ssl=ssl_context)
        async with aiohttp.ClientSession(connector=connector) as session:
            async with session.post(
                device_endpoint,
                data={
                    "client_id": self.client_id,
                    "scope": " ".join(self._scopes()),
                },
            ) as resp:
                if resp.status != 200:
                    body = await resp.text()
                    raise click.ClickException(f"Device authorization request failed (HTTP {resp.status}): {body}")
                device_data = await resp.json()

        device_code = device_data["device_code"]
        user_code = device_data["user_code"]
        verification_uri = device_data.get("verification_uri_complete") or device_data["verification_uri"]
        expires_in = device_data.get("expires_in", 600)
        interval = device_data.get("interval", DEVICE_FLOW_POLL_INTERVAL)

        print(
            f"\nOpen the following URL in your browser and enter the code:"
            f"\n  {verification_uri}"
            f"\n  Code: {user_code}\n"
        )

        token_endpoint = config["token_endpoint"]
        deadline = time.monotonic() + expires_in

        while time.monotonic() < deadline:
            await anyio.sleep(interval)

            client = self.client()
            try:
                token = await run_sync(
                    lambda c=client: c.fetch_token(
                        token_endpoint,
                        grant_type=DEVICE_FLOW_GRANT_TYPE,
                        device_code=device_code,
                    )
                )
                return token
            except Exception as e:
                error_msg = str(e)
                if "authorization_pending" in error_msg:
                    continue
                if "slow_down" in error_msg:
                    interval += 5
                    continue
                if "expired_token" in error_msg or "expired" in error_msg:
                    raise click.ClickException("Device code expired. Please try again.") from e
                if "access_denied" in error_msg:
                    raise click.ClickException("Authorization was denied by the user.") from e
                raise click.ClickException(f"Device flow token request failed: {e}") from e

        raise click.ClickException("Device code expired (timeout). Please try again.")


def should_use_device_flow(device_flow_flag: bool | None = None) -> bool:
    if device_flow_flag is not None:
        return device_flow_flag
    if os.environ.get(JMP_OIDC_DEVICE_FLOW, "").strip() == "1":
        return True
    if os.environ.get("VSCODE_INJECTION", "").strip() == "1":
        return True
    return False


def decode_jwt(token: str):
    try:
        return json.loads(extract_compact(token.encode()).payload)
    except (ValueError, KeyError, TypeError) as e:
        raise ValueError(f"Invalid JWT format: {e}") from e


def decode_jwt_issuer(token: str):
    return decode_jwt(token).get("iss")


def get_token_expiry(token: str) -> int | None:
    """Get token expiry timestamp (Unix epoch seconds) from JWT.

    Returns None if token doesn't have an exp claim.
    """
    return decode_jwt(token).get("exp")


def get_token_remaining_seconds(token: str) -> float | None:
    """Get seconds remaining until token expires.

    Returns:
        Positive value if token is still valid
        Negative value if token is expired (magnitude = how long ago)
        None if token doesn't have an exp claim
    """
    exp = get_token_expiry(token)
    if exp is None:
        return None
    return exp - time.time()


# Token expiry warning threshold in seconds (5 minutes)
TOKEN_EXPIRY_WARNING_SECONDS = 300



def format_duration(seconds: float) -> str:
    """Format a duration in seconds as a human-readable string.

    Args:
        seconds: Duration in seconds (can be negative for past times)

    Returns:
        Formatted string like "2h 30m" or "5m 10s"
    """
    abs_seconds = abs(seconds)
    hours = int(abs_seconds // 3600)
    mins = int((abs_seconds % 3600) // 60)
    secs = int(abs_seconds % 60)

    if hours > 0:
        return f"{hours}h {mins}m"
    if mins > 0:
        return f"{mins}m {secs}s"
    return f"{secs}s"

File: rust/jumpstarter-cli/src/oidc.rs
//! Minimal OIDC client (`jumpstarter_cli_common/oidc.py`). Currently implements
//! discovery + the refresh-token grant used by `auth refresh` and `login`; the
//! authorization-code / token-exchange / password grants land with `login`.

use serde::Deserialize;

/// Build a reqwest client honoring `--insecure-tls` and, like Python's
/// `_get_ssl_context`, the `SSL_CERT_FILE` env var (its PEM is added as an extra
/// trust root so HTTPS to a custom-CA endpoint works without `-k`).
pub fn build_http_client(insecure: bool) -> Result<reqwest::Client, String> {
    let mut builder = reqwest::Client::builder().danger_accept_invalid_certs(insecure);
    if !insecure {
        if let Ok(path) = std::env::var("SSL_CERT_FILE") {
            if let Ok(pem) = std::fs::read(&path) {
                for cert in reqwest::Certificate::from_pem_bundle(&pem).unwrap_or_default() {
                    builder = builder.add_root_certificate(cert);
                }
            }
        }
    }
    builder.build().map_err(|e| e.to_string())
}

/// OIDC client parameters (`oidc.Config`).
pub struct OidcConfig {
    pub issuer: String,
    pub client_id: String,
    pub insecure_tls: bool,
    pub offline_access: bool,
}

/// The discovery document fields the CLI uses.
#[derive(Debug, Deserialize)]
pub struct Discovery {
    // Used by the login authorization-code grant.
    #[allow(dead_code)]
    #[serde(default)]
    pub authorization_endpoint: Option<String>,
    pub token_endpoint: String,
}

/// A token endpoint response.
#[derive(Debug, Deserialize)]
pub struct Tokens {
    pub access_token: String,
    #[serde(default)]
    pub refresh_token: Option<String>,
    // Surfaced by the token-exchange grant in login.
    #[allow(dead_code)]
    #[serde(default)]
    pub id_token: Option<String>,
}

impl OidcConfig {
    pub fn new(issuer: impl Into<String>, client_id: impl Into<String>) -> Self {
        Self {
            issuer: issuer.into(),
            client_id: client_id.into(),
            insecure_tls: false,
            offline_access: false,
        }
    }

    fn http_client(&self) -> Result<reqwest::Client, String> {
        build_http_client(self.insecure_tls)
    }

    /// Base scopes plus `offline_access` when requested (`oidc.py:62-77`).
    pub fn scopes(&self) -> Vec<&'static str> {
        let mut scopes = vec!["openid", "profile"];
        if self.offline_access {
            scopes.push("offline_access");
        }
        scopes
    }

    /// `GET {issuer}/.well-known/openid-configuration`.
    pub async fn discover(&self) -> Result<Discovery, String> {
        let url = format!(
            "{}/.well-known/openid-configuration",
            self.issuer.trim_end_matches('/')
        );
        let resp = self
            .http_client()?
            .get(&url)
            .send()
            .await
            .map_err(|e| format!("OIDC discovery failed: {e}"))?;
        if !resp.status().is_success() {
            return Err(format!(
                "OIDC discovery at {url} returned HTTP {}",
                resp.status()
            ));
        }
        resp.json()
            .await
            .map_err(|e| format!("invalid OIDC discovery document: {e}"))
    }

    /// POST a grant to the token endpoint and parse the response.
    async fn fetch_token(
        &self,
        token_endpoint: &str,
        params: &[(&str, &str)],
    ) -> Result<Tokens, String> {
        let resp = self
            .http_client()?
            .post(token_endpoint)
            .form(params)
            .send()
            .await
            .map_err(|e| format!("token request failed: {e}"))?;
        if !resp.status().is_success() {
            let status = resp.status();
            let body = resp.text().await.unwrap_or_default();
            return Err(format!("token endpoint returned HTTP {status}: {body}"));
        }
        resp.json()
            .await
            .map_err(|e| format!("invalid token response: {e}"))
    }

    /// The `refresh_token` grant (`oidc.py:refresh_token_grant`).
    pub async fn refresh_token_grant(&self, refresh_token: &str) -> Result<Tokens, String> {
        let discovery = self.discover().await?;
        let scope = self.scopes().join(" ");
        self.fetch_token(
            &discovery.token_endpoint,
            &[
                ("grant_type", "refresh_token"),
                ("refresh_token", refresh_token),
                ("client_id", &self.client_id),
                ("scope", &scope),
            ],
        )
        .await
    }

    /// The `password` grant (`oidc.py:password_grant`).
    pub async fn password_grant(&self, username: &str, password: &str) -> Result<Tokens, String> {
        let discovery = self.discover().await?;
        let scope = self.scopes().join(" ");
        self.fetch_token(
            &discovery.token_endpoint,
            &[
                ("grant_type", "password"),
                ("username", username),
                ("password", password),
                ("client_id", &self.client_id),
                ("scope", &scope),
            ],
        )
        .await
    }

    /// The RFC 8693 token-exchange grant, exchanging an id_token for an access
    /// token (`oidc.py:token_exchange_grant`). `connector_id` is a Dex extension.
    pub async fn token_exchange_grant(
        &self,
        token: &str,
        connector_id: Option<&str>,
    ) -> Result<Tokens, String> {
        let discovery = self.discover().await?;
        let scope = self.scopes().join(" ");
        let mut params = vec![
            (
                "grant_type",
                "urn:ietf:params:oauth:grant-type:token-exchange",
            ),
            (
                "requested_token_type",
                "urn:ietf:params:oauth:token-type:access_token",
            ),
            (
                "subject_token_type",
                "urn:ietf:params:oauth:token-type:id_token",
            ),
            ("subject_token", token),
            ("client_id", &self.client_id),
            ("scope", &scope),
        ];
        if let Some(cid) = connector_id {
            params.push(("connector_id", cid));
        }
        self.fetch_token(&discovery.token_endpoint, &params).await
    }

    /// The browser authorization-code grant (`oidc.py:authorization_code_grant`):
    /// spin up a localhost callback server, print the authorization URL, await the
    /// redirect, then exchange the code. No PKCE (matching the Python authlib usage).
    pub async fn authorization_code_grant(
        &self,
        callback_port: Option<u16>,
    ) -> Result<Tokens, String> {
        use tokio::io::{AsyncReadExt, AsyncWriteExt};
        use tokio::net::TcpListener;

        let discovery = self.discover().await?;
        let auth_endpoint = discovery
            .authorization_endpoint
            .as_deref()
            .ok_or("OIDC discovery document is missing authorization_endpoint")?;

        let port = resolve_callback_port(callback_port)?;
        let listener = TcpListener::bind(("127.0.0.1", port))
            .await
            .map_err(|e| format!("Failed to start callback server on port {port}: {e}"))?;
        let actual_port = listener.local_addr().map_err(|e| e.to_string())?.port();
        let redirect_uri = format!("http://localhost:{actual_port}/callback");

        let state = random_state();
        let scope = self.scopes().join(" ");
        let auth_url = reqwest::Url::parse_with_params(
            auth_endpoint,
            &[
                ("response_type", "code"),
                ("client_id", self.client_id.as_str()),
                ("redirect_uri", redirect_uri.as_str()),
                ("scope", scope.as_str()),
                ("state", state.as_str()),
            ],
        )
        .map_err(|e| format!("invalid authorization endpoint: {e}"))?;

        // `print("...: ", uri)` renders two spaces before the URL.
        println!("Please open the URL in browser:  {auth_url}");

        let (mut stream, _) = listener
            .accept()
            .await
            .map_err(|e| format!("callback server error: {e}"))?;
        let mut buf = vec![0u8; 16384];
        let n = stream.read(&mut buf).await.map_err(|e| e.to_string())?;
        let request = String::from_utf8_lossy(&buf[..n]);
        let target = request
            .lines()
            .next()
            .and_then(|line| line.split_whitespace().nth(1))
            .unwrap_or("");
        let query = target.split_once('?').map(|(_, q)| q).unwrap_or("");

        let body = "Login successful, you can close this page";
        let response = format!(
            "HTTP/1.1 200 OK\r\nContent-Type: text/plain\r\nContent-Length: {}\r\nConnection: close\r\n\r\n{}",
            body.len(),
            body
        );
        let _ = stream.write_all(response.as_bytes()).await;

        let parsed = reqwest::Url::parse(&format!("http://localhost/?{query}"))
            .map_err(|e| format!("invalid callback URL: {e}"))?;
        let mut code = None;
        let mut got_state = None;
        for (k, v) in parsed.query_pairs() {
            match k.as_ref() {
                "code" => code = Some(v.into_owned()),
                "state" => got_state = Some(v.into_owned()),
                "error" => return Err(format!("authorization failed: {v}")),
                _ => {}
            }
        }
        let code = code.ok_or("callback did not include an authorization code")?;
        if got_state.as_deref() != Some(state.as_str()) {
            return Err("authorization state mismatch (possible CSRF)".to_string());
        }

        self.fetch_token(
            &discovery.token_endpoint,
            &[
                ("grant_type", "authorization_code"),
                ("code", &code),
                ("redirect_uri", &redirect_uri),
                ("client_id", &self.client_id),
            ],
        )
        .await
    }
}

/// Resolve the callback port: explicit flag, else `JMP_OIDC_CALLBACK_PORT`, else 0
/// (OS-assigned) (`oidc.py:127-137`).
fn resolve_callback_port(callback_port: Option<u16>) -> Result<u16, String> {
    if let Some(p) = callback_port {
        return Ok(p);
    }
    match std::env::var(jumpstarter_config::env::JMP_OIDC_CALLBACK_PORT) {
        Err(_) => Ok(0),
        Ok(v) => v.parse::<u16>().map_err(|_| {
            format!(
                "Invalid {} \"{v}\": must be a valid port",
                jumpstarter_config::env::JMP_OIDC_CALLBACK_PORT
            )
        }),
    }
}

/// A random URL-safe state value for CSRF protection.
fn random_state() -> String {
    let mut bytes = [0u8; 16];
    getrandom::getrandom(&mut bytes).expect("OS RNG unavailable");
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}
